src/Lazy/Lazy/ConcurrentLazy.cs

```diff
@@ -0,0 +1,48 @@
+using System;
+
+namespace Lazy;
+
+/// <summary>
+/// Evaluate supplier once, then return previously calculated result. Safe for multithreading.
+/// </summary>
+/// <typeparam name="T"> Type of supplier result. </typeparam>
+public class ConcurrentLazy<T> : ILazy<T>
+{
+    private Func<T> _supplier;
+    private readonly object _lock = new ();
+    private T _result;
+    private volatile bool _isComputed;
+
+    /// <summary>
+    /// Create instance of <see cref="ConcurrentLazy{T}"/>.
+    /// </summary>
+    /// <param name="supplier"> Function to evaluate. </param>
+    /// <exception cref="ArgumentNullException"> Throws exception, when supplier is null. </exception>
+    public ConcurrentLazy(Func<T> supplier)
+    {
+        ArgumentNullException.ThrowIfNull(supplier, $"{nameof(supplier)} cannot be null");
```

Comment:
```suggestion
        ArgumentNullException.ThrowIfNull(supplier);
```
Собственно, ThrowIfNull появился благодаря поддержке атрибута [CallerArgumentExpression] в компиляторе, который позволяет автоматически сделать nameof(supplier).